syntax = "proto2";

message File {
  required string name = 1;
  required string content = 2;
}

message AlphaDAGRequest {
  required bool warning_as_error = 1;
  required bool with_tables = 2;
  required bool with_functions = 3;
  required bool side_effect_first = 4;
  repeated File files = 5;
}

message AlphaDAGResponse {
  repeated string external_required_tables = 1;
  repeated string dag_dot_string = 2;
  optional string error = 3;
}

enum SupportedType {
  STRING = 0;
  INTEGER = 1;
  INT64 = 2;
  BOOLEAN = 3;
  BOOL = 4;
  FLOAT64 = 5;
  FLOAT = 6;
  NUMERIC = 7;
  BYTES = 8;
  TIMESTAMP = 9;
  TIME = 10;
  DATETIME = 11;
  DATE = 12;
  GEOGRAPHY = 13;
  RECORD = 14;
}

enum Mode {
  REPEATED = 0;
  NULLABLE = 1;
  REQUIRED = 2;
}

/* import "google/protobuf/any.proto"; */

message Column {
  required string name = 1;
  required SupportedType type = 2;
  required Mode mode = 3;
  // For record types
  repeated Column fields = 4;
  /* optional string description = 6; */
  /* optional google.protobuf.Any policyTags = 7; */
}

message TableSchema {
  required string table_name = 1;
  repeated Column columns = 2;
}

message AlphaCheckRequest {
  repeated TableSchema external_required_tables_schema = 1;
  required string dag_dot_string = 2;
  repeated File files = 3;
}

message AlphaCheckResponse {
  optional string error = 1;
}

service AlphaSQL {
  // Extract DAG from SQL files
  rpc AlphaDAG(AlphaDAGRequest)
    returns (AlphaDAGResponse) {}

  // Validate DAG
  rpc AlphaCheck(AlphaCheckRequest)
    returns (AlphaCheckResponse) {}
}
